src/librustc_mir/transform/const_prop.rs

```diff
@@ -293,13 +294,20 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         source: MirSource<'tcx>,
     ) -> ConstPropagator<'mir, 'tcx> {
         let def_id = source.def_id();
-        let param_env = tcx.param_env(def_id);
+        let substs = &InternalSubsts::identity_for_item(tcx, def_id);
+        let mut param_env = tcx.param_env(def_id);
+
+        // If we're evaluating inside a monomorphic function, then use `Reveal::All` because
+        // we want to see the same instances that codegen will see. This allows us to `resolve()`
+        // specializations.
+        if !substs.needs_subst() {
+            param_env = param_env.with_reveal_all();
+        }
```

Comment:
I don't understand why this cares about monomorphic functions. `Reveal::All` is orthogonal to being monomorphic, and any assumption tying them together is a bug.

`Reveal::All` is "*literally* anything outside type-checking", and I'm not even sure `Reveal::UserFacing` should even exist, when it was first added we didn't have the polymorphic+specialization-aware `Reveal::All` (based on `needs_substs`) capability.